### ISSUE TITLE
404 on cancel()

### ISSUE THREAD

**Issue by SamPotts** [TTLabs/EvaporateJS#121]
Hey,

I'm using v1.0.0 and am seeing 404's from S3 when hitting the `cancel` method during an upload. A 404 is also returned after the upload is completed too but I guess that's to be expected. I didn't notice this with the previous version.

Evaporate appears to send the `DELETE` request which returns a 204 and then a subsequent GET request which fails and returns:

```xml
<Code>NoSuchUpload</Code>
<Message>The specified upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.</Message>
```

The bucket policy is as follows:
"s3:AbortMultipartUpload",
"s3:GetObjectVersion",
"s3:GetObject",
"s3:ListMultipartUploadParts",
"s3:PutObject"

**Comment by bikeath1337** [comment]
From what you describe, that is not new behavior. If you read the AWS Docs on aborting uploads, it clearly states that one may have to try several times to make sure that the parts have all been aborted. It is not until the last part has been aborted that AWS actually removes the multipart upload. So it is OK to get a 404 because it's an indication that the multipart upload is indeed, gone.

I believe that that 404 you are seeing is actually trapped in the code on line 661. What I discovered was that regardless of how the JS handles the Ajax callback, the console still shows ALL xhr failures and warnings.

**Comment by bikeath1337** [comment]
Here is the link to the docs:

http://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadAbort.html

What I found is that yes, you can query for the parts, but that I needed to make an additional call to check that they were indeed gone. I found that the console reported error did not break the flow of the upload as it was properly handled.

**Comment by SamPotts** [comment]
Yeah, I read those docs earlier but as with AWS docs, it wasn't exactly clear. 

Alright, cool well I guess it's not an issue then if that's the expected behaviour. It's strange they don't use another response code to keep the console clean. A 404 is never a good look :/

**Comment by bikeath1337** [comment]
I hear you but a 404 is actually good here. It's saying that what you want isn't there (any longer)...and that is what you want.

So how is the rest of the build? Should I merge those changes to v1.0.1?

**Comment by SamPotts** [comment]
Yeah, I guess it's just petty anyway. As long as it works. 

Thanks for sorting v1.0.1. Seems to be working fine. The only issue I've come across (other than the ones I mentioned yesterday) is that when I enable the retry option `s3FileCacheHoursAgo` that a completed upload doesn't seem to get removed from local storage and then if the user tries to upload it again (lets face it, users *are* that dumb) it fails. Should it be removed from storage?

**Comment by bikeath1337** [comment]
If you enable the resume logic by setting s3FileCacheHoursAgo, then the file should not be removed from localStorage because it needs to be there to avoid re-uploading what's already present.

If you're seeing an error there, that should be addressed. I just tried the branch this morning and the resume logic worked -- didn't fail -- actually skipped the upload since the file was already on S3.

Let's open a new issue for this, if necessary.

**Comment by SamPotts** [comment]
OK, sounds like we'll wait on using that feature. It's just to a user this would be perceived as a failed upload - which we get enough feedback on as is. If there was a way to determine that the file exists already and let them know, that would be better. 

Is there any garbage collection of the localStorage?

**Comment by bikeath1337** [comment]
This  is where non-face2face communication breaks down. I do not know what you are trying to say. Let's schedule a Skype or hangout.

**Comment by tomsaffell** [comment]
>If there was a way to determine that the file exists already and let them know, that would be better

@bikeath1337  If I'm understanding right, @SamPotts is suggesting that the complete callback should include a message that the file 'upload' succeeded because it had in fact already been uploaded. He can then use that to message his users in his UI, so they dont have the confusing experience of selecting a large file, and then seconds later seeing that it is 100% uploaded. (OK, that wouldn't be confusing in _many parts of the world_, or the bits of the US that have Google Fiber, but it would be confusing for the rest of us)

I agree that the 404 for the 'did all the parts really get deleted' check is appropriate. It's ugly to see it in the console, but given S3's design, it does seem that the only way to verify that the parts arent there is to try them, and then get a 404. To the best of my knowledge XHR response status has zero effect of JS execution - i.e. the operation of EvaporateJS is not negatively affected by the 404.

**Comment by bikeath1337** [comment]
That's what I have been trying to say: it's supposed to do that and if it doesn't, it's a bug. I am not a fan of long back/forth in chat because a quick call or Skype is a much higher bandwidth method...the demonstration is that I'm still not clear if we do or do not have a bug. If we have one, I want to fix it now :)

I know that on our production version, we do treat already uploaded files with success.

**Comment by tomsaffell** [comment]
@bikeath1337 

My suggestion is that
>the complete callback should *include a message* that the file 'upload' succeeded because it had in fact *already been uploaded*

AFAICT, it's not doing that, because the .complete callback used for the 'already there' case is identical to the .complete used when a file is actually uploaded and then 'completed' on S3, no?

https://github.com/TTLabs/EvaporateJS/blob/master/evaporate.js#L619
https://github.com/TTLabs/EvaporateJS/blob/master/evaporate.js#L585

**Comment by bikeath1337** [comment]
The method signature of the complete method changed
That must be it
I thought I documented it in the change log

--
Bobby
646-321-6011

**Comment by SamPotts** [comment]
I think the issue lies with our setup in that in some cases we specify a filename because of issues in the past with special characters that are allowed in some filenames (french with ' etc) that the usual URI encoding didn't help with. So in cases where the filename isn't customer facing we change the name and then when uploading to S3 we can specify the path of course. The resulting issue is then that the key created based on the original blobs filename. So of course the filename is in storage the second time around but the check fails. I'll have to try creating a new File before sending to Evaporate with the new filename I think. 

I think sometimes it's better to have the discussion documented for others in case they have similar issues?

**Comment by bikeath1337** [comment]
Makes sense . That is our use case, too. For that reason, the complete callback now includes the s3 object key so you can know if it reused what was uploaded or simply uploaded with the key you specified.

**Comment by SamPotts** [comment]
So to reproduce:
- Upload file 
- File uploads fine and the following is added to storage:
```javascript
"{"Kishi_Bashi_-_It_All_Began_With_a_Burst.mp3-audio/mp3-2016-01-10T04:27:29.000Z-9272977":{"awsKey":"uploads/qmbyppsxd1w/media-cover-53959.mp3","uploadId":"-removed-","fileSize":9272977,"fileType":"audio/mp3","lastModifiedDate":"2016-01-10T04:27:29.000Z","partSize":6291456,"signParams":{-removed-},"createdAt":"2016-03-09T23:30:49.707Z","completedAt":"2016-03-09T23:31:29.824Z","eTag":"\"dca43c2c2a79b4c0b705121c3994e7b0-2\""}}"
```
- Verify file exists
- Attempt to upload file again 
- Console shows the intial GET fails, returning the NoSuchFile error above. When removing the query parameters though, the file exists. 
- Appears to try and upload the file but fails

Here's the console:
https://cloudup.com/cqP31yG43eC

**Comment by bikeath1337** [comment]
Thanks Sam
Will try later

--
Bobby
646-321-6011

**Comment by SamPotts** [comment]
Awesome. Let me know if you need anything more from me :+1:

**Comment by bikeath1337** [comment]
Can you 

1. send me your EvaporateJS configuration
2. awsUploads (pretty-printed)? 
3. tell me if the entire file was successfully uploaded and you are trying to not re-upload at all?
4. or a part of the file was uploaded and you are trying to complete it?

If the entire file is uploaded, then you must enable the s3GetObject permission for the IAM user on that bucket.

Here is what I see when I try to upload a file that's already on s3:

```text
evaporate..js?body=1:156 add
evaporate..js?body=1:241 processQueue   length: 1
evaporate..js?body=1:269 starting FileUpload 0
evaporate..js?body=1:227 onFileUploadStatusChange
evaporate.js?body=1:241 processQueue   length: 1
evaporate.js?body=1:600 headObject
evaporate.js?body=1:1051 setupRequest() Object {method: "HEAD", path: "/bikeath1337-evaporate/Button%20to%20Left.m4v-715589659", step: "head_object"}auth: "GIwDGulSpLC8oPWEurJysqT226E="dateString: "Thu, 10 Mar 2016 11:57:36 GMT"method: "HEAD"on200: (xhr)onErr: ()onFailedAuth: (xhr)onGotAuth: ()path: "/bikeath1337-evaporate/Button%20to%20Left.m4v-715589659"step: "head_object"x_amz_headers: Object__proto__: Object
evaporate.js?body=1:1147 authorizedSend() head_object
evaporate.js?body=1:1172 authorizedSend got signature for step: 'head_object'    sig: GIwDGulSpLC8oPWEurJysqT226E=
evaporate.js?body=1:618 headObject found matching object on S3.
evaporate.js?body=1:227 onFileUploadStatusChange
evaporate.js?body=1:241 processQueue   length: 1
JSON.stringify(JSON.parse(localStorage.awsUploads), null, 2)
"{
  "Button to Left.m4v-video/mp4-2016-02-24T23:22:51.000Z-4781827": {
    "awsKey": "Button%20to%20Left.m4v-715589659",
    "uploadId": "fL58fimeBqCCfus4LJjcA0MIZ4HmCN553M3Rje0hHKGSAXYdtPdL9k2sZC4k834piqBFSvUh_EaTM.MxnvkdiWJdXrSSDboX.K_sO.6NudxZNh.9E5ilsMOBtaSijHTO",
    "fileSize": 4781827,
    "fileType": "video/mp4",
    "lastModifiedDate": "2016-02-24T23:22:51.000Z",
    "partSize": 6291456,
    "createdAt": "2016-03-10T11:39:02.231Z",
    "firstMd5Digest": "kORSSdOUyrOJOh+SbkInPg==",
    "completedAt": "2016-03-10T11:39:10.484Z",
    "eTag": "\"cc986e7c83c9071a183401a5efc13bf9-1\""
  }
}"
```

**Comment by SamPotts** [comment]
1. 
```javascript
evaporate = new Evaporate({
	signerUrl:          	        (config.useTempBucket ? config.urls.sign_temp : config.urls.sign_digital),
	timeUrl: 				config.urls.time,
	aws_key:            	        config.aws[config.uploader].Key,
	bucket:             	        config.aws[config.uploader].Bucket,
	logging:            	        engine.config.debug || false,
	encodeFilename:     	false,
	progressIntervalMS: 	100,
	s3FileCacheHoursAgo: 	1
});
```

2. 
```javascript
"{
    "Kishi_Bashi_-_It_All_Began_With_a_Burst.mp3-audio/mp3-2016-01-10T04:27:29.000Z-9272977": {
        "awsKey": "uploads/qtyhdgfatwy/media-cover-29727.mp3",
        "uploadId": "eg04BgALGRsrUQqda4w8j1SlcFvQuzL6jQ5Q2lBdDbwNBMlu3zgmrEMQ1bucjvIJOKWtH5dfOHNYZm0nEryIlQW4VfxE.64cmD0wV_ojMWPXBmOFLsAkolyDHNylowbC",
        "fileSize": 9272977,
        "fileType": "audio/mp3",
        "lastModifiedDate": "2016-01-10T04:27:29.000Z",
        "partSize": 6291456,
        "signParams": {
            "token": "IjXVT8COhDLWKgmVw1xtD8I35GD5YRXSAnpACB5v7c0%3d"
        },
        "createdAt": "2016-03-10T12:14:23.810Z"
    }
}"
```

3. Entire file verified as before - re-upload causes the issue I mentioned

4. Resuming a partial upload seems to work fine. It's literally after complete and re-attempting upload (it happens) that causes the issue

**Comment by SamPotts** [comment]
Ahhh, I think maybe it's the `allowS3ExistenceOptimization` option is required. So essentially you need to set that to true if you use the `s3FileCacheHoursAgo` option? I wasn't aware of the dependency.

**Comment by SamPotts** [comment]
Added the MD5 options and everything seemed to work - complete callback was fired and no errors. I used the SparkMD5 example. I think we're getting somewhere.

**Comment by bikeath1337** [comment]
That setting was added later for flexibility reasons...we can adjust it and document it better. Thanks for the information. Let's think about how best to make this work smoothly. I don't think we're too far off.

**Comment by SamPotts** [comment]
Adding the MD5 options seemed to be what solved it for me. The difference between my output to console and yours was the `HEAD` request so I dug into the code and found this: 

```javascript
if (con.allowS3ExistenceOptimization && me.firstMd5Digest === md5_digest) {
    headObject(awsKey);
```

Which led me to try the MD5 and the `allowS3ExistenceOptimization` options. That seemed to be the only usage of `headObject()`.

**Comment by bikeath1337** [comment]
I haven't had time to think about this today. It strikes me though, that we need to adjust how these params interact so that we avoid confusion. For example, I assumed you were using md5 digests since that is sort of the most amazing things of this version of EvaporateJS. 

I do think that there is an implicit link between keeping track of uploads and the ability to be confident that the upload to be re-used was indeed copied with fidelity. Maybe that relationship needs to be clarified or addressed.

**Comment by SamPotts** [comment]
Well, we'll keep the MD5 digests in there for sure. There's some really awesome stuff in v1.0.0 - really good work :+1:

**Comment by bikeath1337** [comment]
Well, I didn't mean it that way. What I meant was that the caching logic isn't meaningful if you can't guarantee that what you cached is actually accurate. So maybe cache is disabled without md5 digests?

**Comment by SamPotts** [comment]
That sounds like a good idea to me, yeah. Not like it's much extra overhead to add the md5 lib and it sounds like it's worth it, given the benefits.
